webpack/assets/javascripts/react_app/routes/common/reducerHOC/withDataReducer.js

```diff
@@ -0,0 +1,28 @@
+import Immutable from 'seamless-immutable';
+
+const initialState = Immutable({
+  isLoading: true,
+  hasError: false,
+  message: { type: 'empty', text: '' },
+});
+
+const withDataReducer = controller => (
+  state = initialState,
+  { type, payload }
+) => {
+  switch (type) {
+    case `${controller}_PAGE_DATA_RESOLVED`:
+      return state.merge(payload);
+
+    case `${controller}_PAGE_DATA_FAILED`:
+      return state.merge(payload);
```

Comment:
I think the consumer should fire `PAGE_HIDE_LOADING` separately, to give more freedom.
regarding `hasError` (and your other comments):
usually the payload will have `{ hasError: true, message: { type, text }}`, I thought about making the reducers more opinionated but why not make the state extendable? yes it will put the responsibility to change those fields on the consumer but I think its healthy for developers, so `state.merge` gives freedom and responsibility.

Reply:
I don't mind `state.merge` itself. But in the case of failing to load, we already know that `hasError` should likely be true, so we don't really need every place using this HOC to send it along - unless for some reason it should not be true. Or we could get rid of the duplicate actions and do `${controller}_PAGE_DATA_DONE` - if they perform the same action. Essentially this is duplicating the error state both in the payload and in the action.

Reply:
I agree on the `hasError`, updated.
`isLoading` should be fully controlled by the consumer imo